### ISSUE TITLE
[Meteor 1.8.1] data not removed from client due to ddp-server internal changes

### ISSUE THREAD

**Issue by wildhart** [peerlibrary/meteor-reactive-publish#40]
In Meteor 1.8.1-beta.3 `ddp-server` was changed internally to use `Map` and `Set` instead of `Object`. This resulted in `meteor-reactive-publish` failing to remove data from the client which no longer matches the publication's query.  See https://github.com/meteor/meteor/issues/10447 for an explanation and repro.

`meteor-reactive-publish` needs to be updated to accommodate this new structure, while maintaining backwards compatibility.

**Comment by etyp** [comment]
@mitar any idea if this is something you're planning to change in the coming weeks? I know it's super frustrating that it's coming from a change to Meteor core 👎 

We'd be happy to help try to do an initial PR, I'm just not sure how you'd prefer to handle backwards-compatibility with objects instead of just `Set` and `Map`.

**Comment by mitar** [comment]
I am currently writing my PhD thesis and for the next 2 months I will not have time to look into this. Sorry.

I would just make wrap all access with:

```js
if (this.something instanceof Set) {
    this.something.set(...);
}
else {
   this.something[...] = ...;
}
```

**Comment by brucejo75** [comment]
Ha!  Thanks for ferreting this out!  Was digging into this for the past day as I just upgraded to `1.8.1`.
I assume that this likely affects a lot of packages that `reactive-publish` depends on as well?

So... it seems like the workaround is to rip out `reactive-publish`?

**Comment by mitar** [comment]
Or make a PR.

**Comment by brucejo75** [comment]
Fair point.  I would be happy to help on that front and I could make it a priority.

But I am handicapped on a couple of fronts 😢:

1. coffeescript is practically impenetrable for me.  I have repaired some problems with a cofeescripted package previously.  I can do a couple of things, but it is a lot of work for me to understand the code.
2. Debugging coffeescript is quite tedious when I don't know the language.  console does not accept `@foobar` even though it is right there in the cofeescript.
3. I may be wrong, but I think fixes likely need to occur in packages that `reactive-publish` depends on too.  And I assume this is more coffeescript.

If this was vanilla js, then I would be willing to hop in and might do a creditable job.  But coffeescript really kills my productivity.

**Comment by etyp** [comment]
We're working to get a PR together now. Hopefully it can be done soon as we don't want to disrupt mitar during thesis. Will update here soon.

**Comment by mitar** [comment]
I am open to converting it to vanilla JS in the future. (But I would not have time to review it currently.)

**Comment by etyp** [comment]
Hats off to @zodern for opening the PR: https://github.com/peerlibrary/meteor-reactive-publish/pull/42

@mitar is this something you'd have time to review at all?

**Comment by mitar** [comment]
This looks good. Can somebody else test it as well please? Once I get one more confirmation I can merge this and release.

**Comment by etyp** [comment]
@wildhart would you be able to test this as well? See #42 

It's working for me and @brucejo75 but no harm in having others test.

**Comment by wildhart** [comment]
Works for me!  Tested with Meteor 1.8.1-beta.17 and 1.8.1.

Thanks @zodern!

**Comment by mitar** [comment]
Fixed in #42.
